core/src/main/java/io/grpc/ServiceDescriptor.java

```diff
@@ -67,4 +69,39 @@ public String getName() {
   public Collection<MethodDescriptor<?, ?>> getMethods() {
     return methods;
   }
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof ServiceDescriptor)) {
+      return false;
+    }
+    ServiceDescriptor that = (ServiceDescriptor) other;
+    if (!name.equals(that.name)) {
+      return false;
+    }
+    // Fast path, have same order.
+    if (methods.equals(that.methods)) {
+      return true;
+    }
+    // Super slow path, the price of doing tricky things.  Don't be tricky.
+    List<MethodDescriptor<?, ?>> thisMethodsCopy = new ArrayList<MethodDescriptor<?, ?>>(methods);
+    List<MethodDescriptor<?, ?>> thatMethodsCopy =
+        new ArrayList<MethodDescriptor<?, ?>>(that.methods);
+    Comparator<MethodDescriptor<?, ?>> compare = new Comparator<MethodDescriptor<?, ?>>() {
+      @Override
+      public int compare(MethodDescriptor<?, ?> o1, MethodDescriptor<?, ?> o2) {
+        // MethodDescriptors are currently reference-equal only.  This also happens to handle the
+        // case where there are duplicate MethodDescriptors present.
+        return System.identityHashCode(o1) - System.identityHashCode(o2);
+      }
+    };
+    Collections.sort(thisMethodsCopy, compare);
+    Collections.sort(thatMethodsCopy, compare);
+    return thisMethodsCopy.equals(thatMethodsCopy);
```

Comment:
If there are true duplicates, then it seems safe to throw them away. A much bigger problem would be if there are two _different_ method descriptors for the same method name, but things would go badly with that anyway.

Wow, there really isn't any validation, including no verification that the methods are in the service. We should probably move or copy the validation in ServerServiceDefinition here.